language/src/rustspec_to_coq_ssprove_pure.rs

```diff
@@ -0,0 +1,1748 @@
+                    .group()
+                    .nest(2),
+            )
+            .append(RcDoc::line_())
+            .append(RcDoc::as_string(")"))
+            .group(),
+    }
+}
+
+fn make_list<'a, I: IntoIterator<Item = RcDoc<'a, ()>>>(args: I) -> RcDoc<'a, ()> {
+    RcDoc::as_string("[")
+        .append(
+            RcDoc::line_()
+                .append(RcDoc::intersperse(
+                    args.into_iter(),
+                    RcDoc::as_string(";").append(RcDoc::line()),
+                ))
+                .group()
+                .nest(2),
+        )
+        .append(RcDoc::line_())
+        .append(RcDoc::as_string("]"))
+        .group()
+}
+
+fn make_typ_tuple<'a, I: IntoIterator<Item = RcDoc<'a, ()>>>(args: I) -> RcDoc<'a, ()> {
+    RcDoc::as_string("(")
+        .append(
+            RcDoc::line_()
+                .append(RcDoc::intersperse(
+                    args.into_iter(),
+                    RcDoc::space()
+                        .append(RcDoc::as_string("'×"))
+                        .append(RcDoc::line()),
+                ))
+                .group()
+                .nest(2),
+        )
+        .append(RcDoc::line_())
+        .append(RcDoc::as_string(")"))
+        .group()
+}
+
+fn make_paren<'a>(e: RcDoc<'a, ()>) -> RcDoc<'a, ()> {
+    RcDoc::as_string("(")
+        .append(RcDoc::softline_().append(e).group().nest(2))
+        .append(RcDoc::as_string(")"))
+        .group()
+}
+
+fn translate_constructor<'a>(enum_name: TopLevelIdent) -> RcDoc<'a> {
+    RcDoc::as_string(enum_name.string)
+}
+
+fn translate_enum_name<'a>(enum_name: TopLevelIdent) -> RcDoc<'a> {
+    translate_toplevel_ident(enum_name)
+}
+
+fn translate_enum_case_name<'a>(
+    enum_name: BaseTyp,
+    case_name: TopLevelIdent,
+    explicit: bool,
+) -> RcDoc<'a> {
+    match enum_name {
+        BaseTyp::Named(name, opts) => match opts {
+            None => translate_constructor(case_name),
+            Some(tyvec) => if explicit && tyvec.len() != 0 {
+                RcDoc::as_string("@")
+            } else {
+                RcDoc::nil()
+            }
+            .append(translate_constructor(case_name))
+            .append(
+                if (name.0).string == "Option" || (name.0).string == "Result" {
+                    RcDoc::nil()
+                } else {
+                    RcDoc::as_string("(")
+                        .append(translate_toplevel_ident(name.0))
+                        .append(RcDoc::as_string(")"))
+                },
+            )
+            .append(if explicit && tyvec.len() != 0 {
+                RcDoc::space().append(RcDoc::intersperse(
+                    tyvec.into_iter().map(|(x, _)| translate_base_typ(x)),
+                    RcDoc::space(),
+                ))
+            } else {
+                RcDoc::nil()
+            }),
+        },
+        _ => panic!("should not happen"),
+    }
+}
+
+fn translate_base_typ<'a>(tau: BaseTyp) -> RcDoc<'a, ()> {
+    match tau {
```

Comment:
You have the base file already and there's a lot of duplicate code here. Can you factor out more to re-use more code for all the coq backends?